[Feature] Auxiliary Storage

pblottiere opened the pull request:
## Description

This PR adds the Auxiliary Storage feature discussed in the QEP qgis/QGIS-Enhancement-Proposals#27.

A new tab is available in vector layer properties to manage auxiliary storage :  

![as323](https://user-images.githubusercontent.com/9266424/29836108-180924ee-8cec-11e7-8845-91436b9477cd.png)


Moreover a new action *Store data in the project* is available in the data defined menu providing an easy way to manage auxiliary data for a property : 

![as00](https://user-images.githubusercontent.com/9266424/29835982-9d85468a-8ceb-11e7-90cc-2bd28f441c9a.png)

Auxiliary data is stored in a sqlite database and managed thanks to the OGR data provider (instead of the spatialite provider) to keep as small as possible the database file. This database file (with extension *.qgd*) is either saved just next to the project file or directly embedded within the new *.qgz* format.

Units tests will come soon.

I look forward to your reactions and suggestions! Tank you.

Warning: there's some known issues which may disturb while testing this feature:
- https://issues.qgis.org/issues/17029
- https://issues.qgis.org/issues/17068

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

pblottiere committed:
[needs-doc] Add auxiliary storage tab

pblottiere committed:
Add auxiliary storage mechanism in project

pblottiere committed:
Auxiliary layers may be created or loaded from auxiliary storage

pblottiere committed:
[needs-doc] Auxiliary layers can be created from the auxiliary storage tab

pblottiere committed:
Add an origin (PAL or Diagram) to QgsProperty

pblottiere committed:
[needs-doc] Add a new button in property menu to create auxiliary fields

pblottiere committed:
Implement action create auxiliary field

pblottiere committed:
Update auxiliary storage tab and data defined buttons according to auxiliary fields

pblottiere committed:
Update map tools to work with auxiliary fields in not editable mode

pblottiere committed:
In case of a null variant, the default value is returned (otherwise moved labels/diagrams are not displayed on opening when show/hide DD are NULL)

pblottiere committed:
Do not show join configuration for auxiliary layer

pblottiere committed:
Implement clear action

pblottiere committed:
Implement delete action

pblottiere committed:
Implement export action

pblottiere committed:
Implement delete field action

pblottiere committed:
Rowid column from ogr provider is ignored thanks to a blacklist join option

pblottiere committed:
Update documentation and sip binding

pblottiere committed:
Add a safe guard

pblottiere committed:
Button to store data in project activates the property if it exists in auxiliary layer

pblottiere committed:
Fix enable state for change label map tool

nyalldawson commented:
Argh - sorry about the malformed comments above. Was reviewing on my phone, and it messed up the review.

Unanswered question #1:
- what happens/should happen when a vector layer with auxiliary storage is duplicated/cloned?

nyalldawson commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -29,6 +30,60 @@
+     * \returns def The property definition
+     * \returns joined The join prefix is tok into account if true
+     */
+    static QString name( const QgsPropertyDefinition &def, bool joined = false );
```
This is messy - why not just rename this method as nameFromProperty and avoid the using QgsField::name requirement?

pblottiere replied:
You're right, done!

nyalldawson commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -29,6 +30,60 @@
+     * \param def Definition of the property to be stored by this auxiliary
+     *  field.
+     */
+    QgsAuxiliaryField( const QgsPropertyDefinition &def );
```
explicit?

pblottiere replied:
Done

nyalldawson commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -71,6 +126,25 @@ class CORE_EXPORT QgsAuxiliaryLayer : public QgsVectorLayer
     QgsVectorLayerJoinInfo joinInfo() const;
 
     /**
+     * Returns true if the property is stored in the layer yet, false
```
yet -> already

pblottiere replied:
done

nyalldawson commented on src/core/qgsauxiliarystorage.cpp:
```diff
@@ -19,13 +19,129 @@
+
+QgsAuxiliaryField::QgsAuxiliaryField( const QgsField &f )
+{
+  const QStringList parts = f.name().split( '_' );
```
This constructor looks a bit fragile to me - is it required anywhere? Or could we just force use of QgsAuxiliaryField( const QgsPropertyDefinition &def )?

pblottiere replied:
Indeed it's pretty fragile, that's why this constructor is private. Only QgsAuxiliaryLayer may use it (friend class).

It's currently used in QgsAuxiliaryLayer::auxiliaryFields method.

nyalldawson commented on src/core/qgsauxiliarystorage.cpp:
```diff
@@ -19,13 +19,129 @@
+      case QgsPropertyDefinition::DataTypeBoolean:
+        type = QVariant::Int; // sqlite does not have a bool type
+        break;
+      default:
```
Remove this default - it's better to get the compiler warning when new types are added

pblottiere replied:
done

nyalldawson commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -29,6 +30,60 @@
+ *
+ * \since QGIS 3.0
+ */
+class CORE_EXPORT QgsAuxiliaryField : public QgsField
```
I'm not that comfortable with subclassing QgsField here. From my interpretation this class is only used as a convenience to creating QgsFields with the correct types/names, and after creation isn't used anywhere. In that case it could just be reduced to a static `QgsField createAuxiliaryField( const QgsPropertyDefinition &def)` method which returns a new QgsField with the correct type/name.

pblottiere replied:
It's also used to retrieve the underlying QgsPropertyDefinition in QgsVectorLayerProperties to build the widget tree in auxiliary storage tab (property name and origin).

nyalldawson replied:
But couldn't that be done by re-parsing the field name in a static method too?

pblottiere replied:
Yes it's possible to re-parse the field name. 

I think it's more elegant to inherit from QgsField to encapsulate this logic. Moreover, it could be useful in the future if we want to add some more specific behavior to auxiliary field.

However, you're the Jedi here, so if you think it's better to remove QgsAuxiliaryField and use static methods instead, just tell me :)

nyalldawson replied:
Let's get some more opinions :) @m-kuhn @wonder-sk?

m-kuhn replied:
With the inheritance approach, is there never a problem of slicing?

```
QgsAuxiliaryField field;
fields.append(field);
fields.last(); // will return a QgsField, not a QgsAuxiliaryField
```

nyalldawson commented on src/app/qgslabelinggui.cpp:
```diff
@@ -611,6 +611,15 @@ void QgsLabelingGui::updateUi()
     chkMergeLines->setToolTip( QString() );
   }
+
+  mCoordXDDBtn->updateFieldLists();
```
Why are these calls to updateFieldLists() required? The field list is only used in the popup menu, and updateFieldList is already called before the menu is shown

pblottiere replied:
The field list is also used in updateGui to properly update the icon.

nyalldawson commented on src/app/qgsvectorlayerproperties.cpp:
```diff
@@ -1470,6 +1472,33 @@ void QgsVectorLayerProperties::updateAuxiliaryStoragePage()
+        QTreeWidgetItem *item = new QTreeWidgetItem();
+
+        if ( prop.origin() == QgsPropertyDefinition::Pal )
+          item->setText( 0, "Pal" );
```
This should be "labeling". PAL is an internal name for the label placement engine, it shouldn't be exposed anywhere to users

pblottiere replied:
done

nyalldawson commented on src/app/qgsmaptoollabel.cpp:
```diff
@@ -707,3 +792,19 @@ QgsMapToolLabel::LabelDetails::LabelDetails( const QgsLabelPosition &p )
   }
 }
+
+bool QgsMapToolLabel::isAuxiliaryField( QgsVectorLayer *layer, int index ) const
```
I think this should be moved to QgsVectorLayer itself - it could be handy elsewhere too

pblottiere replied:
Good point! Done.

nyalldawson commented on src/app/qgsvectorlayerproperties.h:
```diff
@@ -207,6 +210,7 @@ class APP_EXPORT QgsVectorLayerProperties : public QgsOptionsDialogBase, private
 
     QAction *mAuxiliaryLayerActionNew;
     QAction *mAuxiliaryLayerActionClear;
+    QAction *mAuxiliaryLayerActionDelete;
```
Initialise pointers in the header

pblottiere replied:
done

nyalldawson commented on src/core/qgsauxiliarystorage.cpp:
```diff
@@ -168,6 +168,7 @@ QgsAuxiliaryLayer::QgsAuxiliaryLayer( const QString &pkField, const QString &fil
   mJoinInfo.setEditable( true );
   mJoinInfo.setUpsertOnEdit( true );
   mJoinInfo.setCascadedDelete( true );
+  mJoinInfo.setJoinFieldNamesBlackList( QStringList() << "rowid" ); // introduced by ogr provider
```
QStringLiteral ("rowid")

pblottiere replied:
done

nyalldawson commented:
Lovely stuff! This is all very much appreciated :tada: 

Looking over the original QEP, is this part still planned too? "The creation of such auxiliary fields will also be transparent for the user: the manual labeling tools will be available and if no fields already exist for data-defined properties, they will be automatically created through auxiliary fields."

pblottiere committed:
Init bool value in header and do not init qstring attributes with empty value

pblottiere commented:
> Why's this restricted to spatial layers?

@haubourg

haubourg commented:
>Looking over the original QEP, is this part still planned too? "The creation of such auxiliary fields will also be transparent for the user: the manual labeling tools will be available and if no fields already exist for data-defined properties, they will be automatically created through auxiliary fields."

@nyalldawson in fact, after testing I asked Paul to remove that. Opening a project was automatically adding tables to each layer (spatial or not) and that looked intrusive, and potentially slow.

I also found that it's always hard to teach users to first play with automatic placement settings when it's too easy to manually move labels. So I was wondering if making it too easy wasn't ruining efforts of trainers or GIS administrators somewhere

My idea to make it straight simple is now to create an assistant somewhere to automatically set all labeling widgets, add styles with geometry generators for label callouts, with multiple style choices for rendering / fonts and so on. In one word, we now have the core features to easily port EasyCustomLabeling plugin :). We had that in separated QEP about connectors, custom paths for plugins, but the guy funding that quit his job for any reason... (me indeed).

pblottiere committed:
Expand documentation for QgsAuxiliaryStorage constructors

nyalldawson commented:
> in fact, after testing I asked Paul to remove that. Opening a project was automatically adding tables to each layer (spatial or not) and that looked intrusive, and potentially slow.

Couldn't we just add these when they are first required? So e.g. always have the label tools available, but only create the auxiliary field when a label in that layer is first moved.

I really liked this idea - it was bringing us to feature parity with one of the few remaining areas MapInfo is ahead of us. I think even an having an assistant to automatically set this up is too hard for many users to understand.

pblottiere commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -0,0 +1,334 @@
+     * \param copy Parameter indicating if a copy of the database has to be used
+     */
+    QgsAuxiliaryStorage( const QString &filename = QString(), bool copy = true );
+
```
> @nyalldawson : It's a bit unclear here what happens when filename is empty, and how it relates to copy. E.g. mention that if filename is empty and copy is true a temporary file is constructed. And what happens if filename is empty but copy is false?

I updated the documentation, let me know if it suits you.

haubourg commented:
@nyalldawson agreed for Mapinfo fully.  From a user point of view  I'm still unsure it is a good idea to have pinned labels so easily and then turn the map into a fixed scale map. And just moving / rotating labels in current map tools is not enough to reach mapinfo, we need callouts and alignment settings also, which is a lot of data defined settings to do. That's why I was thinking of an assistant to help users do that easily. 
From a developper point of view, current  maptools are not "active layer" aware, they might need some refactor to trigger the auxiliary field creation on the fly. I just had a look at the rotate Label (which is broken due to rotation changes ) and I'm afraid users will have random "auxiliary data creation" dialogs popping when they pick labels from other layers. 
We might probably think a maptool refactor  before doing that? 
In fact, we face something similar with the new node tool, not active layer focused any more. We have now very heterogeneous ergonomics with active layer and maptools.

haubourg commented:
@nyalldawson so I'm not saying it's not possible but I'd like to be sure this is really what we want.

pblottiere committed:
Fix saveAs

pblottiere committed:
Replace reset/release by std::move for unique_ptr

pblottiere committed:
Expand documentation for QgsAuxiliaryLayer class

pblottiere committed:
Fix pointer initalization

pblottiere committed:
Update sip binding

pblottiere committed:
Always remove join info before resetting auxiliary layer

pblottiere committed:
Fix comment

pblottiere committed:
Origin in QgsProperty is a string and pal is replaced by labeling

pblottiere committed:
Fix documentation

pblottiere committed:
Add QStringLiteral

pblottiere committed:
Initialise pointers in header for QActions

pblottiere committed:
Move isAuxiliaryField to QgsVectorLayer

pblottiere commented:
> what happens/should happen when a vector layer with auxiliary storage is duplicated/cloned?

For now, the auxiliary layer is not kept in the clone but  I don't have a string opinion here. 

But if we want to clone the auxiliary layer too, then the underlying sqlite table should be copied to avoid some conflicts.

pblottiere committed:
Remove default in switch/case to rise warning if a new data type is added

pblottiere committed:
Add explicit keyword

pblottiere committed:
Rename name method to nameFromProperty

pblottiere commented:
@nyalldawson Firstly, thank you for your extensive review! I think I addressed all comments and questions. Let me know what you think about!

And by the way, what do you ( @nyalldawson and @haubourg ) decide about the map tool stuff?

nyalldawson commented:
I've just given a test run. Here's some thoughts:

- please don't forget about the "Change Label" toolbar action. It's currently disabled when the layer isn't editable, unlike the move/rotate label actions. But it should be treated the same as these.
- i need to manually hit "apply" in the style dock to get the properties to apply - this should be automatic
- do we want to hide auxilliary fields from the attribute table and identify results? (i'd say yes - they could default to the "hidden" editor widget, but be aware that there's a bug where hidden fields are being shown in the attribute table - i can't find the issue report for this, but it's there somewhere)
- should users be allowed to add their own auxiliary fields? i can see that there's potentially use cases for this...

nyalldawson commented on src/core/qgsproperty.h:
```diff
@@ -124,7 +117,7 @@ class CORE_EXPORT QgsPropertyDefinition
      * \param origin The origin of the property
      */
-    QgsPropertyDefinition( const QString &name, const QString &description, StandardPropertyTemplate type, Origin origin = QgsPropertyDefinition::Pal );
+    QgsPropertyDefinition( const QString &name, const QString &description, StandardPropertyTemplate type, const QString &origin = QStringLiteral( "labeling" ) );
```
origin should default to QString(), not labeling.

Also note that we had issues in the past when using QStringLiteral in header files, so better to avoid that.

pblottiere replied:
done

nyalldawson commented:
> > what happens/should happen when a vector layer with auxiliary storage is duplicated/cloned?
> For now, the auxiliary layer is not kept in the clone but I don't have a strong opinion here.

>But if we want to clone the auxiliary layer too, then the underlying sqlite table should be copied to avoid some conflicts.

Agreed - a clone should duplicate the existing sqlite table, then update the clone to use the new table but retain all existing links for data defined properties (pointing to new table).

nyalldawson commented:
@haubourg 

> agreed for Mapinfo fully. From a user point of view I'm still unsure it is a good idea to have pinned labels so easily and then turn the map into a fixed scale map.

I disagree. If a user wants to turn a map into a fixed scale map, let's make it easy for them. I used to manage a MapInfo rollout with ~400 non-GIS users (non techy people, too), yet they were totally comfortable with moving labels around in MapInfo and would frequently do this to improve their maps. This is the benchmark we need to hit here.

In QGIS it's easy to unfreeze labels anyway, and the ability to click and drag to unfreeze multiple at one would make this quick to do if needed. (incidentally, seems there's a bug in that freeze tool - it's not automatically refreshed on the map).

>  And just moving / rotating labels in current map tools is not enough to reach mapinfo, we need callouts and alignment settings also, which is a lot of data defined settings to do. That's why I was thinking of an assistant to help users do that easily.

True, but that's a different topic/function. Let's make label moving/rotating/hiding/custom styling as easy as possible first, given that it's being tackled in this PR :)

> From a developper point of view, current maptools are not "active layer" aware, they might need some refactor to trigger the auxiliary field creation on the fly. I just had a look at the rotate Label (which is broken due to rotation changes ) and I'm afraid users will have random "auxiliary data creation" dialogs popping when they pick labels from other layers.

Again, I don't think that's a big problem. The "auxiliary data creation" dialog could be improved a bit so that it's not so technical and scary for users, but in the end if they WANT to move a label which is in a layer which doesn't currently have aux data attached, the dialog is really just making this easy for them to do that. I see it as QGIS saying "hey, you're wanting to move this label. That's great! You've just got to answer this one question for me, and then you I'll let you move and tweak those labels however you want." (what a friendly little fellow QGIS is!)

pblottiere committed:
Remove default value for origin of property

haubourg commented:
@nyalldawson pretty convincing :+1:  
@pblottiere So let's activate the label maptools by default and create what's needed on the fly when required. Tell us if the state of the maptools requires more work than expected.

nyalldawson commented:
> @nyalldawson pretty convincing :+1:

![cheerydependablearcticseal-small](https://user-images.githubusercontent.com/1829991/29949648-949a7f80-8ef9-11e7-85a1-7d25b1da088a.gif)

One thing that occurred to me last night - we also have some map tools for working with data defined symbol properties (rotate point symbol tool, offset point symbol tool). Should these also be addressed?

nyalldawson commented:
> do we want to hide auxilliary fields from the attribute table and identify results? (i'd say yes - they could default to the "hidden" editor widget, but be aware that there's a bug where hidden fields are being shown in the attribute table - i can't find the issue report for this, but it's there somewhere)

I thought some more on this - I think what should happen is:

- aux field created which maps to any of the predefined label tools (e.g. x/y/rotation/those settings available through the label properties tool) should have their editor widget set to hidden by default
- fields mapped to other properties should not be set to hidden (otherwise there's no way to edit the values for these). E.g. if i bind the "background shape radius" to a aux field, that field should be shown in the attribute table/identify results.
- as a follow up to this, any fields bound to a QgsPropertyDefinition::ColorWithAlpha or ColorNoAlpha property should have their editor widget set to the color widget by default

Thoughts?

m-kuhn commented:
> do we want to hide auxilliary fields from the attribute table and identify results? (i'd say yes - they could default to the "hidden" editor widget, but be aware that there's a bug where hidden fields are being shown in the attribute table - i can't find the issue report for this, but it's there somewhere)

visibility of fields in the attribute table is no longer controlled by hidden, it's controlled by "organize fields".

haubourg commented:
> I thought some more on this - I think what should happen is:
> 
> - aux field created which maps to any of the predefined label tools (e.g. x/y/rotation/those settings available through the label properties tool) should have their editor widget set to hidden by default

+1

> - fields mapped to other properties should not be set to hidden (otherwise there's no way to edit the values for these). E.g. if i bind the "background shape radius" to a aux field, that field should be shown in the attribute table/identify results.

+1. Good catch, I only focused on labeling tests.

> - as a follow up to this, any fields bound to a QgsPropertyDefinition::ColorWithAlpha or ColorNoAlpha property should have their editor widget set to the color widget by default
> Thoughts?

of course, this would be better if easy to do.

haubourg commented:
> One thing that occurred to me last night - we also have some map tools for working with data defined symbol properties (rotate point symbol tool, offset point symbol tool). Should these also be addressed?

I'm not totally convinced here. Editing toolbars strongly suggest you are editing main data source.I would let the current "Store data in the project" checkbox do the job in that case.

pblottiere committed:
Move/rotate/hide/pin map tools are always available

pblottiere commented:
> So let's activate the label maptools by default and create what's needed on the fly when required. 

Done but it still remains some work to do on *Change Label* map tool. I'm taking a look.

m-kuhn commented on src/core/qgsauxiliarystorage.h:
```diff
@@ -197,6 +199,30 @@ class CORE_EXPORT QgsAuxiliaryLayer : public QgsVectorLayer
+     *
+     * \returns The index of the auxiliary field or -1
+     */
+    static int createProperty( QgsPalLayerSettings::Property p, const QString &providerId, QgsVectorLayer *vlayer );
```
Please prefer longer, descriptive names

```py
QgsAuxiliaryLayer.createProperty(property=QgsPalLayerSettings.Bold)
```

reads much better than


```py
QgsAuxiliaryLayer.createProperty(p=QgsPalLayerSettings.Bold)
```

pblottiere replied:
Oops sorry, it's fixed

pblottiere committed:
Change label maptool is always available and underlying properties are created

pblottiere commented:
> please don't forget about the "Change Label" toolbar action. It's currently disabled when the layer isn't editable, unlike the move/rotate label actions. But it should be treated the same as these.

Done. I just noticed that max/min scale data defined properties are just took into account after reloading the project... I don't know why (yet).

nyalldawson commented:
>>One thing that occurred to me last night - we also have some map tools for working with data defined symbol properties (rotate point symbol tool, offset point symbol tool). Should these also be addressed?
>I'm not totally convinced here. Editing toolbars strongly suggest you are editing main data source.I would let the current "Store data in the project" checkbox do the job in that case.

That's fine by me - but I don't think it's handled in this PR yet (creating aux fields for symbology data defined properties)

nyalldawson commented:
> visibility of fields in the attribute table is no longer controlled by hidden, it's controlled by "organize fields".

Is this predictable behavior though? It seems odd to me that the fields won't show in form view, but will in the table view.

haubourg commented:
>That's fine by me - but I don't think it's handled in this PR yet (creating aux fields for symbology data defined properties)

True, We have the "store data " checkbox not doing anything currently outside of Labeling fields. 
@pblottiere do yo have to handle by hand every auxiliary field name or is that something automated?

m-kuhn commented:
> Is this predictable behavior though? It seems odd to me that the fields won't show in form view, but will in the table view.

It could be adjusted that when a widget type is changed to hidden it will also be deactivated for the attribute table. And when it's changed from hidden to something else, that it's set to visible. This would leave maximum configurability available while keeping it easy to configure for the 90% default cases.

pblottiere committed:
Auxiliary fields for symbol layers

pblottiere commented:
> That's fine by me - but I don't think it's handled in this PR yet (creating aux fields for symbology data defined properties)

It's done (but I needed a non const vector layer in QgsSymbolLayerWidget to be able to create the auxiliary field)

pblottiere committed:
Some cleanup in maptool

pblottiere committed:
Hide auxiliary columns which can be edited by "change label properties" map tool

nyalldawson commented:
@nirvn can I request you to checkout this branch and give a test run? You've always got good UX insights, so I'm keen to hear your thoughts on how this behaves.

nyalldawson commented on src/core/qgsauxiliarystorage.cpp:
```diff
@@ -0,0 +1,659 @@
+  return rc;
+}
+
+int QgsAuxiliaryLayer::createProperty( QgsPalLayerSettings::Property property, const QString &providerId, QgsVectorLayer *layer )
```
I think this specific use method belongs in QgsPalLayerSettings, and the diagram version in QgsDiagramLayerSettings. It seems messy to me to have usage specific functions in QgsAuxiliaryLayer.

pblottiere replied:
Are you sure? Because it's pretty specific to the auxiliary storage stuff considering that the aim is to create a new field in the auxiliary layer. However it's true that the new property is then activated in layer settings... Maybe the method is badly named and leads to confusion?

nyalldawson commented:
@m-kuhn 
> It could be adjusted that when a widget type is changed to hidden it will also be deactivated for the attribute table. And when it's changed from hidden to something else, that it's set to visible. This would leave maximum configurability available while keeping it easy to configure for the 90% default cases.

+1 to this

nyalldawson commented:
Latest test results:

1. If I click the "label properties" tool, and click on a label for the first time for a layer, I get the "set primary key" dialog. After clicking here I'd expect the label properties dialog to show, but instead I get nothing and need to reclick the label.

2. Being able to change the font style of a single label without any config or table changes is killer. Thank you SO much for this!

3. Note to self: add a "show buffer" checkbox to the label properties dialog so that buffers can be enabled/disabled for individual labels too.

4. There's no effect if I choose "store data in the project" for point symbol rotation or size (at the symbol level, not symbol layer level)

5. I see "Store data in the project" options for property override buttons where they should not be. E.g. composer. Selecting the option here crashes.

6. The approach used to hide the aux fields is incorrect. They should just auto set the widget type to "hidden" initially, but leave the option for users to later change this if desired. The current approach prevents them from ever being shown.

m-kuhn commented:
Sorry in advance if this question has been answered before, it's just a random thought.

Is it possible to "promote" an auxiliary field to a provider field? I imagine the situation where I start working on an isolated "branch" of a project to bring it back to the production "master" data source later on.

pblottiere commented:
> The approach used to hide the aux fields is incorrect. They should just auto set the widget type to "hidden" initially, but leave the option for users to later change this if desired. The current approach prevents them from ever being shown.

I probably didn't understand what you mean but if you click on the *Organize columns* button, you have the next window where auxiliary widgets may be displayed if necessary:

![organize](https://user-images.githubusercontent.com/9266424/29993867-bff42c8c-8fba-11e7-9d03-6940fcfa43d7.png)

m-kuhn commented:
> > It could be adjusted that when a widget type is changed to hidden it will also be deactivated for the attribute table. And when it's changed from hidden to something else, that it's set to visible. This would leave maximum configurability available while keeping it easy to configure for the 90% default cases.
> 
> +1 to this

Should this only happen when it's done manually (vector layer properties) or also when setting it through the API?
For both I can imagine situations where it will be unexpected.

pblottiere committed:
Automatically update the editor widget setup according to the standard template of the property newly created

pblottiere commented:
> as a follow up to this, any fields bound to a QgsPropertyDefinition::ColorWithAlpha or ColorNoAlpha property should have their editor widget set to the color widget by default

It's now automatically done when a new auxiliary field is created.

pblottiere committed:
A simple click is enough to create auxiliary layer and display "change label properties" window

pblottiere commented:
> If I click the "label properties" tool, and click on a label for the first time for a layer, I get the "set primary key" dialog. After clicking here I'd expect the label properties dialog to show, but instead I get nothing and need to reclick the label.

It was desired but you're right, it's much better this way :). Done.

pblottiere committed:
Fix auxiliary field creation for QgsSymbolListWidget

pblottiere commented:
> There's no effect if I choose "store data in the project" for point symbol rotation or size (at the symbol level, not symbol layer level)

It's fixed.

haubourg commented:
> Sorry in advance if this question has been answered before, it's just a random thought.
> 
> Is it possible to "promote" an auxiliary field to a provider field? I imagine the situation where I start working on an isolated "branch" of a project to bring it back to the production "master" data source later on.

Hi Matthias, this was not in the list, no. Users might use the field calculator, waiting for something more integrated. 
An additional button could be shown in the 'auxiliary storage' tab of layer properties.

pblottiere committed:
Do not activate auxiliary data button everywhere

pblottiere commented:
> I see "Store data in the project" options for property override buttons where they should not be. E.g. composer.

"Store data in the project" action is now only available in : 
- QgsDiagramProperties
- QgsLabelingGui
- QgsSymbolLayerWidget
- QgsSymbolsListWidget

pblottiere committed:
Auxiliary field creation is correctly managed for width in QgsSymbolsListWidget

pblottiere committed:
Auxiliary data is cloned too

pblottiere commented:
> > For now, the auxiliary layer is not kept in the clone but I don't have a strong opinion here.
> >But if we want to clone the auxiliary layer too, then the underlying sqlite table should be copied to avoid some conflicts.

> Agreed - a clone should duplicate the existing sqlite table, then update the clone to use the new table but retain all existing links for data defined properties (pointing to new table).

An auxiliary layer is cloned too during the duplicate action. However, auxiliary data has to be committed to have a proper duplicated table. I added a note in the documentation about this point.

nyalldawson commented:
>> The approach used to hide the aux fields is incorrect. They should just auto set the widget type to "hidden" initially, but leave the option for users to later change this if desired. The current approach prevents them from ever being shown.
> I probably didn't understand what you mean but if you click on the Organize columns button, you have the next window where auxiliary widgets may be displayed if necessary:

You're correct, sorry. I was expecting the widget type to be set to "hidden" as well. Do you think this should also be done?

I also noticed that:
1. move a label using the tool
2. x/y fields are hidden from attribute table automatically (=good)
3. manually show them again
4. add another aux field to the layer
5. x/y fields have been hidden again (=bad)

pblottiere committed:
Set editor widget setup to Hidden

pblottiere commented:
> You're correct, sorry. I was expecting the widget type to be set to "hidden" as well. Do you think this should also be done?

Yes you're right, it's done.  

> I also noticed that:
>1. move a label using the tool
>2. x/y fields are hidden from attribute table automatically (=good)
>3. manually show them again
>4. add another aux field to the layer
>5. x/y fields have been hidden again (=bad)

It's fixed.

pblottiere committed:
Update documentation and sip binding

pblottiere committed:
Add a safe guard

nyalldawson commented:
> Warning: there's some known issues which may disturb while testing this feature:
> https://issues.qgis.org/issues/17068

Fixed in https://github.com/qgis/QGIS/pull/5118

pblottiere committed:
Revert a modification in diagram provider

pblottiere committed:
Fix join information in clone

pblottiere commented:
Travis build is finally green!

Regarding the previous comments, there's still some points to address:
- keeping or not keeping *QgsAuxiliaryField* class?
- promote" an auxiliary field to a provider field? (maybe in another PR?)
- add unit tests

nyalldawson commented:
I think also:
- decide whether the restriction to spatial layers is valid
- possible allow manual creation of auxiliary layers? (e.g. users can manually create auxilliary fields if they have extra data they want to save in a project)

This is looking really excellent to me now! Before merge I'd love to have some testing by @nirvn and also I think for a change of this magnitude a second opinion review on the code would be valuable (@wonder-sk @m-kuhn @elpaso?)

nirvn commented:
@pblottiere , I've had a quick look at this PR. Fantastic work.

I did spot an issue with the current implementation whereas auxiliary storage for symbol layer properties (i.e., size, color, etc.) doesn't support multiple symbol layers. 

For e.g., if I have a marker's symbol composed of two symbol layers (lets say two circle markers), it is not possible to use two distinct auxiliary storage fields for both sizes of these two layers via "store data in the project" menu item.

haubourg commented:
Mmm, good catch Mathieu. We designed this at start for labeling where we have no nested properties. @pblottiere do you think just incrementing the aux field name with a sequence ' _xx ' when needed could work?

pblottiere commented:
> do you think just incrementing the aux field name with a sequence ' _xx ' when needed could work?

Nope. Because before displaying the *Store data in the project* button, I have to detect if an auxiliary field already exists for the current property. To do that, I need a unique way to identify my auxiliary field and for now, the origin (diagram, labeling, symbology) and the name of the property are used. 

But in the case of marker's symbol composed of two symbol layers, it's not enough. So I need another parameter to have a unique way for identifying my auxiliary field...

I take a look.

pblottiere commented:
@nyalldawson  Could we consider providing a unique id to identify a QgsSymbolLayer instance? 

Thus an auxiliary field for a symbol layer could be identified thanks to its origin, the property name and the id of the symbol layer.

nirvn commented:
@haubourg , @pblottiere , this is not only an issue for multi-layer symbols, is it not also a need for rule-based labelling?

haubourg commented:
> @haubourg , @pblottiere , this is not only an issue for multi-layer symbols, is it not also a need for rule-based labelling?

True, for every potential nested config, ruled based labeling too then

pblottiere commented:
> this is not only an issue for multi-layer symbols, is it not also a need for rule-based labelling?

I think that the underlying mechanism for rule-based labelling is also multiple QgsSymbolLayer instances. So if we have a unique way to identify these instances, it's a win for both cases.

Unless I missed something?

pblottiere committed:
Auxiliary storage can be used on non spatial layer

pblottiere commented:
> decide whether the restriction to spatial layers is valid

I removed this restriction.

pblottiere commented:
> possible allow manual creation of auxiliary layers? (e.g. users can manually create auxilliary fields if they have extra data they want to save in a project)

+1 but maybe in another PR? Because this one is already big enough I think :)

pblottiere committed:
Properly update layer settings when an auxiliary field is deleted

pblottiere committed:
Add a custom name for symbol layers in property definition

pblottiere committed:
Allow user to create its own auxiliary field

pblottiere committed:
Remove QgsAuxiliaryField class

pblottiere commented:
> I did spot an issue with the current implementation whereas auxiliary storage for symbol layer properties (i.e., size, color, etc.) doesn't support multiple symbol layers.

> For e.g., if I have a marker's symbol composed of two symbol layers (lets say two circle markers), it is not possible to use two distinct auxiliary storage fields for both sizes of these two layers via "store data in the project" menu item.

@nirvn For symbol layers, a window is opened when the user click on *Store data in the project*. In this window, a name has to be indicated and is used as a complement to *origin* and *property name*. Thus we have a unique way to retrieve data defined for symbol layers. 

And thank you for having tested this PR!

pblottiere commented:
@nyalldawson 

> I'm not that comfortable with subclassing QgsField here

I removed *QgsAuxiliaryField* class. 

> possible allow manual creation of auxiliary layers?

I added a new button to do that :  

![add](https://user-images.githubusercontent.com/9266424/30089867-6e83d272-92a7-11e7-860f-009b0cceb97e.png)

pblottiere commented:
I think I took into account all suggestions so I just have to write some unit tests now.

@nyalldawson Thank you a lot for your involvement in this work!

pblottiere committed:
Update documentation and sip binding

nirvn commented:
Fantastic work @pblottiere .

m-kuhn commented on src/ui/qgsvectorlayerpropertiesbase.ui:
```diff
@@ -1908,6 +1917,216 @@ border-radius: 2px;</string>
+            <item row="0" column="0">
+             <widget class="QLabel" name="label_6">
+              <property name="text">
+               <string>Auxiliary Storage tables can contain additional data that should only belong to the project file. For instance, specific location or rotation for labels. Auxiliary data are saved in qgd files (sqlite). New fields can be added from any data defined widget when needed. Be aware that this information will NOT be saved in you datasource but only in the project file.</string>
```
"Auxiliary Storage tables" or "Auxiliary storage tables"?

pblottiere replied:
I updated the label accordingly

m-kuhn commented:
Awesome stuff @pblottiere I currently don't have the time to review it, but I feel quite comfortable after all the work that has gone into this. Plus there is still some time to fix things before release.
Looking very much forward to this :)

pblottiere committed:
Update information label

pblottiere commented:
@m-kuhn I restarted the build several times but I get every time the next error on Travis:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

What must I do?

m-kuhn commented:
Rebase on master, this was just fixed an hour ago, travis changed something.

pblottiere commented:
> Rebase on master, this was just fixed an hour ago, travis changed something.

OK I did that. Thank you for the fast answer ;)

pblottiere committed:
Remove Q_FOREACH

pblottiere committed:
Actions are already initialized in header

pblottiere committed:
Some minor refactoring

pblottiere committed:
Add tests

pblottiere commented:
@m-kuhn @nyalldawson Do want to take a last review or something? Because I would like your approval before merging a PR of this magnitude...

haubourg commented:
Afraid of the red nuclear button? :rocket:

pblottiere committed:
Update sip binding

pblottiere commented:
> Afraid of the red nuclear button?

Yes a little :). 

I'll merge this PR at the end of the day if I've no more comments in the meantime.

nyalldawson commented:
> I'll merge this PR at the end of the day if I've no more comments in the meantime.

Can we please hold off on that until we're sure no-one else can review? I'd still appreciate a 'second opinion' review for work with this importance.

pblottiere commented:
> Can we please hold off on that until we're sure no-one else can review? I'd still appreciate a 'second opinion' review for work with this importance.

Yes of course! I leave you the final decision to merge this PR so :).

And thank you again @nyalldawson for your involvement in this work!

m-kuhn commented on src/core/qgsproperty.h:
```diff
@@ -125,25 +127,57 @@ class CORE_EXPORT QgsPropertyDefinition
     QString name() const { return mName; }
 
     /**
+     * Sets the name of the property
```
Is the `\since` test still not working?

nyalldawson replied:
The test only works for classes, not members. And in any case this class was added in 3.0, so the \since isn't required here.

m-kuhn replied:
Oh, I always thought it works on member level and wondered how the test finds out if it was added in a different version than the class (somehow would need access to the current branches target version)

m-kuhn commented on src/core/qgsproperty.h:
```diff
@@ -125,25 +127,57 @@ class CORE_EXPORT QgsPropertyDefinition
+    void setName( const QString &name ) { mName = name; }
+
+    /**
+     * Returns the origin of the property
```
Would be great if you could add a tiny bit more information here (explanation or examples)

pblottiere replied:
Indeed. Done.

m-kuhn commented on src/core/qgsvectorlayerjoininfo.h:
```diff
@@ -131,6 +131,36 @@ class CORE_EXPORT QgsVectorLayerJoinInfo
+     *
+     * \since QGIS 3.0
+     */
+    void setJoinFieldNamesBlackList( const QStringList &blackList ) { mBlackList = blackList; }
```
What's this blacklist used for?
Aren't joins able to ignore certain fields already? I was just thinking "never say never" ;)

pblottiere replied:
Yes it's currently possible to ignore certain fields with the subset mechanism. However, it's completely static. 

If you have a joined layer with `fieldA` and `fieldB`, you may define a subset to only see `fieldA` from the target layer and ignore `fieldB`.

Then, if you add some fields in the joined layer, like `fieldC`, the target layer won't see it as `fieldC` is not in the subset.

However, it's constraining in our case because we add on the fly new columns in the joined layer (the auxiliary layer) when a new property has to be managed. 

So I added a blacklist mechanism which is fully dynamic. You just have to give some field to ignore and then, even if some fields are added in the joined layer, then the target layer will see them.

m-kuhn replied:
Will it be possible for a user to unblacklist fields? Or is this something we really don't want someone to do?

pblottiere replied:
I asked myself the same question. As I wasn't sure, the blacklist option is not available in ui for users...

pblottiere committed:
Expand documentation

nyalldawson commented:
@pblottiere 

I gave a quick test to the latest version of this PR. Some comments (sorry!):

- When I tested I accidentally used a project with a filtered layer. I hadn't thought of this before, but things behave quite odd in this scenario. Trying to modify the filter results in QGIS continually warning that it's not supported, and that the layer needs to be converted to a virtual layer. This message should be disabled when the joined field is an auxiliary field.

- Whenever a label/symbol property is bound to a aux field, the layer should be emitting styleChanged() in order for the style dock to update correctly. That's currently not being done and the style dock settings can be out of sync with the actual layer's settings (e.g. by moving a label - the data defined x/y buttons aren't updated to show they are now bound to a field)

- the aux fields aren't working correctly with rule based labeling. e.g. add two label rules, each with x/y stored in the project = things go weird when you try to move labels.

- i was messing with aux fields on a two layer marker symbol, and obtained this crash: https://gist.github.com/nyalldawson/30746cc2ea3bd349b97d5921a0b6401e

haubourg commented:
@nyalldawson 
Hi, Paul is on vacations for two weeks (he already postponed them enough before). 
We'll see that when he's back. 

> > When I tested I accidentally used a project with a filtered layer. I hadn't thought of this before, but things behave quite odd in this scenario. Trying to modify the filter results in QGIS continually warning that it's not supported, and that the layer needs to be converted to a virtual layer. This message should be disabled when the joined field is an auxiliary field.

Good catch!

> Whenever a label/symbol property is bound to a aux field, the layer should be emitting styleChanged() in order for the style dock to update correctly. That's currently not being done and the style dock settings can be out of sync with the actual layer's settings (e.g. by moving a label - the data defined x/y buttons aren't updated to show they are now bound to a field)

The styling dock issue was already identified, thanks for finding the culprit.

> the aux fields aren't working correctly with rule based labeling. e.g. add two label rules, each with x/y stored in the project = things go weird when you try to move labels.

Mmm, nasty one, and given the bad state of the labeling maptools, I hope it's not to hard to fix.

haubourg commented:
@nyalldawson , Should we merge now and collect more testing during the two weeks, and in the same time avoid to much rebase workload?  (This is a real question not a gentle push to force the merge)

pblottiere commented:
@nyalldawson 

The style dock seems currently broken (see https://issues.qgis.org/issues/17029)  so I stopped using it during my manual testing on auxiliary layers because I had really some strange behaviors (as you mentioned, style dock settings can be out of sync and this leads to severe underlying issues).

> Whenever a label/symbol property is bound to a aux field, the layer should be emitting styleChanged() in order for the style dock to update correctly. That's currently not being done and the style dock settings can be out of sync with the actual layer's settings (e.g. by moving a label - the data defined x/y buttons aren't updated to show they are now bound to a field)

Indeed, you're right. However, I just tried and it seems that even with the emission of styleChanged(), the style dock is not correctly updated. It's probably related to the same issue...

m-kuhn commented:
Does this also need to be added to src/app/main.cpp:1171?

pblottiere committed:
Check if join information is for auxiliary layer when filtering

pblottiere committed:
Fix checkbox status for auxiliary field

pblottiere committed:
Auxiliary fields are linked to data defined for all subproviders

pblottiere committed:
Ignore archive files given as layername in command line

pblottiere commented:
According to the previous comments, the last thing I have to address is the update of the style dock.

pblottiere committed:
Emit signal to update the style dock

pblottiere commented:
@nyalldawson I took into account your last comments (sorry for the delay). 

If you want to test it, you should remove the previous `.qgd` file(s) because some thing have changed since.

Let me know what you think!

Thank you

nyalldawson commented:
Right, no objections here - fantastic work!

pblottiere commented:
> Right, no objections here

Good news :).

I'll merge this PR tomorrow if no more comments come up in the meantime.

pblottiere committed:
Fix indentation

pblottiere committed:
Update sip binding

m-kuhn commented:
Yay!!
🎉

haubourg commented:
![image](https://user-images.githubusercontent.com/1629853/31329302-b71c86c2-acd9-11e7-8ce1-313efcc5eb9b.png)

haubourg commented:
Bravo Paul.. huge work (and this is not just to move labels.... )

phidrho commented:
Hi @pblottiere,

why is Auxillary storage 'String' field implemented to hold up only 50 characters?
If I understood well it's stored in local SQLite file so why to limit this field?
